src/main/java/org/mastodon/mamut/feature/spot/ellipsoid/SpotEllipsoidFeatureComputer.java

```diff
@@ -0,0 +1,173 @@
+
+	@Parameter
+	private AtomicBoolean forceComputeAll;
+
+	@Parameter
+	private FeatureComputationStatus status;
+
+	@Parameter( type = ItemIO.OUTPUT )
+	private SpotEllipsoidFeature output;
+
+	private String cancelReason;
+
+	@Override
+	public void createOutput()
+	{
+		if ( null == output )
+		{
+			// Try to get output from the FeatureModel, if we deserialized a model.
+			final Feature< ? > feature =
+					model.getFeatureModel().getFeature( SpotEllipsoidFeature.SPOT_ELLIPSOID_FEATURE_SPEC );
+			if ( null != feature )
+			{
+				output = ( SpotEllipsoidFeature ) feature;
+				return;
+			}
+
+			final DoublePropertyMap< Spot > semiAxisA =
+					new DoublePropertyMap<>( model.getGraph().vertices().getRefPool(), Double.NaN );
+			final DoublePropertyMap< Spot > semiAxisB =
+					new DoublePropertyMap<>( model.getGraph().vertices().getRefPool(), Double.NaN );
+			final DoublePropertyMap< Spot > semiAxisC =
+					new DoublePropertyMap<>( model.getGraph().vertices().getRefPool(), Double.NaN );
+			final DoublePropertyMap< Spot > volume =
+					new DoublePropertyMap<>( model.getGraph().vertices().getRefPool(), Double.NaN );
+			// Create a new output.
+			output = new SpotEllipsoidFeature( semiAxisA, semiAxisB, semiAxisC, volume );
+		}
+	}
+
+	@Override
+	public void run()
+	{
+		cancelReason = null;
+		final boolean recomputeAll = forceComputeAll.get();
+
+		if ( recomputeAll )
+		{
+			// Clear all.
+			output.semiAxisA.beforeClearPool();
+			output.semiAxisB.beforeClearPool();
+			output.semiAxisC.beforeClearPool();
+			output.volume.beforeClearPool();
+		}
+
+		final int numTimepoints = bdvData.getNumTimepoints();
+
+		int done = 0;
+
+		final double[][] covarianceMatrix = new double[ 3 ][ 3 ];
+
+		final JamaEigenvalueDecomposition eigenvalueDecomposition = new JamaEigenvalueDecomposition( 3 );
+
+		for ( int timepoint = 0; timepoint < numTimepoints; timepoint++ )
+		{
+
+			status.notifyProgress( ( double ) done++ / numTimepoints );
+
+			final SpatialIndex< Spot > toProcess = model.getSpatioTemporalIndex().getSpatialIndex( timepoint );
+			for ( final Spot spot : toProcess )
+			{
+				if ( isCanceled() )
+					break;
+
+				// Skip if we are not forced to recompute all and if a value is already computed.
+				if ( !recomputeAll && output.semiAxisA.isSet( spot ) )
+					continue;
+
+				spot.getCovariance( covarianceMatrix );
+				eigenvalueDecomposition.decomposeSymmetric( covarianceMatrix );
+				final double[] eigenValues = eigenvalueDecomposition.getRealEigenvalues();
+				double volume = 4d / 3d * Math.PI;
+				for ( int k = 0; k < eigenValues.length; k++ )
+				{
+					final double semiAxis = Math.sqrt( eigenValues[ k ] );
+					volume *= semiAxis;
+					if ( k == 0 )
+						output.semiAxisA.set( spot, semiAxis );
+					else if ( k == 1 )
+						output.semiAxisB.set( spot, semiAxis );
+					else if ( k == 2 )
+						output.semiAxisC.set( spot, semiAxis );
+				}
+				output.volume.set( spot, volume );
+			}
+		}
```

Comment:
I think this code can be simplified a lot. We don't need the three nested loops. The bdvData and mokito are no longer required if you do this change. 
```suggestion
		final int numSpots = model.getGraph().vertices().size();
		for ( Spot spot : model.getGraph().vertices() )
		{
			if( done++ % 1000 == 0) // Maybe it's a good idea to limit overhead, by only update progress every 1000th spot.
				status.notifyProgress( ( double ) done / numSpots );

			if ( isCanceled() )
				break;

			// Skip if we are not forced to recompute all and if a value is already computed.
			if ( !recomputeAll && output.semiAxisA.isSet( spot ) )
				continue;

			spot.getCovariance( covarianceMatrix );
			eigenvalueDecomposition.decomposeSymmetric( covarianceMatrix );
			final double[] eigenValues = eigenvalueDecomposition.getRealEigenvalues();
			double a = Math.sqrt( eigenValues[ 0 ] );
			double b = Math.sqrt( eigenValues[ 1 ] );
			double c = Math.sqrt( eigenValues[ 2 ] );
			double volume = 4d / 3d * Math.PI * a * b * c;
			output.semiAxisA.set( spot, a );
			output.semiAxisB.set( spot, b );
			output.semiAxisC.set( spot, c );
			output.volume.set( spot, volume );
		}
		status.notifyProgress( 1.0 );
```

Reply:
I agree. Changed it.

Reply:
Requires https://github.com/mastodon-sc/mastodon/pull/227 to be merged.

Reply:
It's much better now :+1: